File: backend/src/modules/reporting/handler.ts
import type { AuthenticationMiddleware } from "../../middleware/auth.ts";
import {
  DOCUMENT_TYPE,
  type DocumentType,
} from "../../shared/types/reporting.ts";
import type { IReportingSvc } from "./svc.ts";
import { type Request, type Response, Router } from "express";
import multer from "multer";

export class ReportingHandler {
  private svc: IReportingSvc;
  private authenticationMiddleware: AuthenticationMiddleware;
  public router: Router;
  constructor(
    svc: IReportingSvc,
    authenticationMiddleware: AuthenticationMiddleware,
  ) {
    this.svc = svc;
    this.authenticationMiddleware = authenticationMiddleware;
    this.router = Router();
    this.registerRoutes();
  }

  private registerRoutes = () => {
    this.getParentDocumentByType();
    this.saveParentDocument();
    this.getDocumentURLByFilePath();
    this.getParentDocumentURLByDocumentID();
  };

  private getParentDocumentByType = () => {
    this.router.get(
      "/parents/:parentId/documents/:documentType",
      this.authenticationMiddleware,
      async (req: Request, res: Response) => {
        const { parentId, documentType } = req.params;
        const isValidDocumentType = (
          documentType: string,
        ): documentType is DocumentType => {
          const set = new Set(Object.values(DOCUMENT_TYPE));
          return set.has(documentType as DocumentType);
        };
        if (!isValidDocumentType(documentType)) {
          res.status(400).json({ error: "Invalid document type!" });
          return;
        }
        const { data, error } = await this.svc.getParentDocumentByType(
          parentId,
          documentType,
        );
        if (error) {
          console.error(error);
          res.status(500).json({ error: error.message });
          return;
        }
        res.status(200).json({ data: data });
      },
    );
  };

  private saveParentDocument = () => {
    const upload = multer({ storage: multer.memoryStorage() });
    this.router.post(
      "/parents/:parentId/documents/:documentType",
      this.authenticationMiddleware,
      upload.single("file"),
      async (req: Request, res: Response) => {
        const { parentId, documentType } = req.params;
        const isValidDocumentType = (
          documentType: string,
        ): documentType is DocumentType => {
          const set = new Set(Object.values(DOCUMENT_TYPE));
          return set.has(documentType as DocumentType);
        };
        if (!isValidDocumentType(documentType)) {
          res.status(400).json({ error: "Invalid document type!" });
          return;
        }
        const file = req?.file;
        if (!file) {
          res.status(400).json({ error: "File is required!" });
          return;
        }
        const browserFile = new File([file.buffer], file.originalname, {
          type: file.mimetype,
          lastModified: Date.now(),
        });
        const { data, error } = await this.svc.saveParentDocument(
          parentId,
          documentType,
          browserFile,
        );
        if (error) {
          console.error(error);
          res.status(500).json({ error: error.message });
          return;
        }
        res.status(200).json({ data: data });
      },
    );
  };

  private getDocumentURLByFilePath = () => {
    this.router.get(
      "/documents/*filePath",
      this.authenticationMiddleware,
      async (req: Request, res: Response) => {
        const filePath = Array.isArray(req.params.filePath)
          ? req.params.filePath.join("/")
          : req.params.filePath;
        const { data, error } = await this.svc.getDocumentURLByFilePath(
          filePath,
          "mybucket",
          3600,
        );
        if (error) {
          console.error(error);
          res.status(500).json({ error: error.message });
          return;
        }
        return res.status(200).json({ data: data });
      },
    );
  };

  private getParentDocumentURLByDocumentID = () => {
    this.router.get(
      "/parent-documents/:documentId",
      this.authenticationMiddleware,
      async (req: Request, res: Response) => {
        const { documentId } = req.params;
        const { data, error } =
          await this.svc.getParentDocumentURLByDocumentID(documentId);
        if (error) {
          console.error(error);
          res.status(500).json({ error: error.message });
          return;
        }
        res.status(200).json({ data: data });
      },
    );
  };
}

File: apps/backend-python/src/modules/reporting/handler.py
from typing import Callable
from uuid import UUID
from fastapi import APIRouter, Response, Depends

from src.modules.reporting.svc import IReportingSvc
from src.shared.types.modules.reporting.enum import DOCUMENT_TYPE
from src.shared.types.modules.reporting.model import ParentDocument
from src.shared.types.response import ApiResponse, AuthMiddlewareResponse, HTTPError


class ReportingHandler:
    def __init__(
        self,
        svc: IReportingSvc,
        auth_middleware: Callable[..., AuthMiddlewareResponse],
    ):
        self.svc = svc
        self.router = APIRouter()
        self.auth_middleware = auth_middleware
        self.register_routes()

    def register_routes(self):
        @self.router.get("/parents/{parent_id}/documents/{document_type}")
        async def get_parent_document_by_type(
            parent_id: UUID,
            document_type,
            response: Response,
            user_result: AuthMiddlewareResponse = Depends(self.auth_middleware),
        ) -> ApiResponse[ParentDocument]:
            if document_type not in DOCUMENT_TYPE:
                return ApiResponse(
                    error=HTTPError(
                        code=400,
                        message="Invalid document type!",
                    )
                )
            user, error = user_result
            if error:
                response.status_code = error.code
                return ApiResponse(error=error)
            data, error = await self.svc.get_parent_document_by_type(
                user_id=parent_id, document_type=document_type
            )
            if error:
                response.status_code = error.code
                return ApiResponse(error=error)
            response.status_code = 200
            return ApiResponse(data=data)

        @self.router.get("/parent-documents/{doc_id}")
        async def get_parent_document_url_by_document_id(
            doc_id: UUID,
            response: Response,
            user_result: AuthMiddlewareResponse = Depends(self.auth_middleware),
        ) -> ApiResponse[str]:
            user, error = user_result
            if error:
                response.status_code = error.code
                return ApiResponse(error=error)
            data, error = await self.svc.get_parent_document_url_by_document_id(
                doc_id=doc_id
            )
            if error:
                response.status_code = error.code
                return ApiResponse(error=error)
            response.status_code = 200
            return ApiResponse(data=data)

        @self.router.get("/documents/{file_path:path}")
        async def get_document_url_by_file_path(
            file_path: str,
            response: Response,
            user_result: AuthMiddlewareResponse = Depends(self.auth_middleware),
        ) -> ApiResponse[str]:
            user, error = user_result
            if error:
                response.status_code = error.code
                return ApiResponse(error=error)
            data, error = await self.svc.get_document_url_by_file_path(path=file_path)
            if error:
                response.status_code = error.code
                return ApiResponse(error=error)
            response.status_code = 200
            return ApiResponse(data=data)
